### PR TITLE
[#1005] Hide individual attachments

### LINKED ISSUE #1005: Allow individual attachments to incoming messages to be hidden
As @RichardTaylor [comments](https://github.com/mysociety/alaveteli/issues/209#issuecomment-2669338) often it is a particular attachment that admins would like to hide. 

In order to enable this, I think we would need to:
- [ ] Add a prominence field to FoiAttachment
- [ ] Add a prominence notes field to FoiAttachment
- [ ] Either update get_attachments_for_display to consider the visibility of the attachments or (in order that we can show them or show the reason why they are hidden) leave it as it is and reference prominence and prominence notes when deciding how to display each attachment
- [ ] As with https://github.com/mysociety/alaveteli/issues/209, we need to make sure that cached versions of attachments are not accessible to users when the attachments are not. 
- [ ] Decide how to display a hidden attachment when it does/doesn't have notes

### PR DESCRIPTION
**Pull request by gbp**
## Relevant issue(s)

Requires https://github.com/mysociety/alaveteli/pull/7324
Fixes #1005

## What does this do?

1. Adds attachment admin UI
2. Adds prominence and reason to `FoiAttachment`
3. Show prominence notice in frontend
4. Add checks for attachments with prominence before downloading/viewed as HTML
5. Prevents attachments with prominence from being indexed by Xapian

## Why was this needed?

Easier admin of responses and attachments.

### Incoming message admin
image

### Attachment admin
image

### Hidden attachment as public / Hidden attachment as requester
image

### Hidden attachment as site admin
image

### Requester only attachment as public
image

### Requester only attachment as requester
image

### Requester only attachment as site admin
image

### Hidden main body attachment as public / Hidden main body attachment as requester
image

### Hidden main body attachment as site admin
image

### Requester only main body attachment as public
image

### Requester only main body attachment as requester
image

### Requester only main body attachment as site admin
image

### CODE REVIEW

File: app/helpers/admin/link_helper.rb [hunk 1]
```diff
@@ -38,6 +38,16 @@ def incoming_message_both_links(incoming_message)
               title: admin_title)
   end
 
+  def foi_attachment_both_links(foi_attachment)
+    title = 'View attachment on public website'
+    icon = eye
+
+    link_to(icon, foi_attachment_path(foi_attachment), title: title) + ' ' +
```
[the remark this answers is not in the document]
Comment (gbp) [reply app/helpers/admin/link_helper.rb:45]: This does link directly to the attachment which is why we have change we're discussing here: https://github.com/mysociety/alaveteli/pull/7324#discussion_r981102135

[the remark this answers is not in the document]
Comment (garethrees) [reply app/helpers/admin/link_helper.rb:45]: Ah, sorry for not being clear.

I appreciate it links to e.g. `/request/foo_bar_baz_2#attachment-77`. What I meant by "directly link to" was `/request/125/response/23/attach/2/fs%2050379341.pdf`

File: app/controllers/admin/foi_attachments_controller.rb [hunk 2]
```diff
@@ -9,11 +9,32 @@ def edit
...
+        old_prominence_reason: @foi_attachment.prominence_reason_previously_was,
+        prominence_reason: @foi_attachment.prominence_reason
+      )
+      @info_request.expire
+
+      flash[:notice] = 'Attachment successfully updated.'
+      redirect_to edit_admin_incoming_message_path(@incoming_message)
+
+    else
+      render action: 'edit'
```
[the remark this answers is not in the document]
Comment (gbp) [reply app/controllers/admin/foi_attachments_controller.rb:28]: We don't do this in the other admin controllers.

[the remark this answers is not in the document]
Comment (garethrees) [reply app/controllers/admin/foi_attachments_controller.rb:28]: I don't think we do when we show model validation errors, but doesn't look like we're doing that here so a fail would be completely silent. I think we only need a `error: 'Save unsuccessful'` flash on the end of the `render` call.

File: app/controllers/admin/foi_attachments_controller.rb [hunk 3]
```diff
@@ -9,11 +9,32 @@ def edit
...
+        editor: admin_current_user,
+        old_prominence: @foi_attachment.prominence_previously_was,
+        prominence: @foi_attachment.prominence,
+        old_prominence_reason: @foi_attachment.prominence_reason_previously_was,
+        prominence_reason: @foi_attachment.prominence_reason
+      )
+      @info_request.expire
+
+      flash[:notice] = 'Attachment successfully updated.'
+      redirect_to edit_admin_incoming_message_path(@incoming_message)
```
[the remark this answers is not in the document]
Comment (gbp) [reply app/controllers/admin/foi_attachments_controller.rb:25]: Re-rendering the form isn't normally done on successful submission, normally we would redirect to the `#show` or `#index` actions but we don't have those, the `incoming_message#edit` action is the closest thing we have. Saying that happy to change if you want.

[the remark this answers is not in the document]
Comment (garethrees) [reply app/controllers/admin/foi_attachments_controller.rb:25]: Yeah, I guess on save I was feeling like I wanted to go back in to `#edit` to check that the record was now as I expected. There's no reason for it not to be – just psychological! Let's leave it as it is for now as it's an easy change if we want to tweak it later.

File: app/controllers/attachments_controller.rb [hunk 4]
```diff
@@ -104,9 +104,13 @@ def authenticate_attachment
...
       return render_hidden
     end
+
     if cannot?(:read, @incoming_message)
       request.format = :html
-      return render_hidden('request/hidden_correspondence')
+      return render_hidden(
+        'request/hidden_correspondence',
+        locals: { message: @incoming_message }
+      )
```
[the remark this answers is not in the document]
Comment (gbp) [reply app/controllers/attachments_controller.rb:113]: I think I missed the fact `render_hidden` renders a template and not a partial. In which case I don't think this is needed. I will revert this commit

File: app/models/incoming_message.rb [hunk 5]
```diff
@@ -498,6 +498,10 @@ def get_attachments_for_display
     return attachments
   end
 
+  def get_attachments_for_search_index
```
[the remark this answers is not in the document]
Comment (gbp) [reply app/models/incoming_message.rb:501]: I certainly agree with this but not in this PR. This method follows the style of other methods in this class which need extracting and renaming.

File: spec/controllers/attachments_controller_spec.rb [hunk 6]
```diff
@@ -597,6 +603,137 @@ def expect_hidden(hidden_template)
...
+            id: info_request.id,
+            part: 2,
+            file_name: 'interesting.pdf',
+            skip_cache: 1
+          }
+      expect(response.media_type).to eq('application/pdf')
+      expect(response).to be_successful
+    end
+
+    it 'should not generate an HTML version of an attachment for a request whose prominence is hidden even for an admin but should return a 404' do
```
[the remark this answers is not in the document]
Comment (gbp) [reply spec/controllers/attachments_controller_spec.rb:723]: Copy-pasted the spec description which is now wrong, the attachment has `hidden` prominence.

File: spec/controllers/attachments_controller_spec.rb [hunk 7]
```diff
@@ -597,6 +603,137 @@ def expect_hidden(hidden_template)
...
```
[the remark this answers is not in the document]
Comment (gbp) [reply spec/controllers/attachments_controller_spec.rb:651]: Copy-pasted the spec description which is now wrong, the attachment has `hidden` prominence.

File: app/views/request/hidden_attachment.html.erb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<%- if !attachment.prominence_reason.blank? %>
```
[the remark this answers is not in the document]
Comment (gbp) [reply app/views/request/hidden_attachment.html.erb]: I think I missed edited this file. It should replicate the `request/hidden_correspondence.html.erb` template.

File: spec/controllers/attachments_controller_spec.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (garethrees) [reply spec/controllers/attachments_controller_spec.rb]: Yeah, I guessed that might be the only issue. My other thought was whether it ought to be testing e.g. hitting a "normal" prominence attachment but the InfoRequest is "hidden". Might be covered elsewhere

**Comment by gbp** [comment]
@garethrees this should be working now. There is loads of fixups do you want to see those or should I go ahead and rebase before you re-review?

**Comment by gbp** [comment]
Specs are green locally not sure why GitHub actions aren't running.

**Comment by garethrees** [comment]
> There is loads of fixups do you want to see those or should I go ahead and rebase before you re-review?

Happy to review as-is.

> Specs are green locally not sure why GitHub actions aren't running.

Might be because of the merge conflict?

**Comment by garethrees** [comment]
On running I get:

Screenshot 2022-10-21 at 15 43 45

Fixed with:

```diff
diff --git a/app/helpers/prominence_helper.rb b/app/helpers/prominence_helper.rb
index 2a26944b4..c81994e99 100644
--- a/app/helpers/prominence_helper.rb
+++ b/app/helpers/prominence_helper.rb
@@ -98,7 +98,7 @@ def default_prominence_reason
     end
   end

-  class InfoRequest::Prominence::Helper < Base # :nodoc:
+  class InfoRequest::Prominence::Helper < ProminenceHelper::Base # :nodoc:
     def user
       prominenceable.user
     end
@@ -166,6 +166,8 @@ def sign_in_notice(*args)

   ::OutgoingMessage::Prominence::Helper = IncomingMessage::Prominence::Helper

+  module FoiAttachment::Prominence ; end
+
   class FoiAttachment::Prominence::Helper < Base # :nodoc:
     def user
       prominenceable.incoming_message.info_request.user
```

File: app/models/ability.rb [hunk 9]
```diff
@@ -42,6 +42,10 @@ def initialize(user, project: nil, public_token: false)
     end
 
     # Viewing messages with prominence
+    can :read, FoiAttachment do |attachment|
+      can_view_with_prominence?(attachment.prominence, attachment.incoming_message.info_request)
```
Comment (garethrees) [inline app/models/ability.rb]: This doesn't cover the case where the `FoiAttachment` is "normal", but it's `IncomingMessage` is reduced prominence. In this situation you can still view the attachment if you know its direct URL (which search engines will). I think it needs to be like:

```ruby
can_view_with_prominence?(attachment.prominence, attachment.incoming_message.info_request) &&
  can?(:read, attachment.incoming_message)
```

This is pretty critical behaviour, so I think it needs tests, somewhere….

**Comment by gbp** [comment]
> On running I get:
> 

Missed a file, pushed a fixup

File: app/models/incoming_message.rb [hunk 10]
```diff
@@ -498,12 +498,6 @@ def get_attachments_for_display
     return attachments
   end
 
-  # rubocop:disable Naming/AccessorMethodName
-  def get_attachments_for_search_index
```
Comment (garethrees) [inline app/models/incoming_message.rb]: Not clear why this is getting removed here?

**Review by garethrees — CHANGES_REQUESTED** [review]
As an admin user I though I should be able to see reduced-prominence main body parts below the prominence reason?

Screenshot 2022-10-21 at 16 25 47

I think this is getting quite a big and complicated PR, so maybe let's pair on this on Monday to get some of the more incidental parts merged, and then get left with just the main body of work that adds attachment prominence.

**Comment by gbp** [comment]
Superseded by #7391

**Pull request closed without merging**